README.md
<b> Bi Directional MQTT to OPCUA Bridge </b>

Expose MQTT topics to an OPCUA server via folders. Supports read/write. 

Usage:

- git clone https://github.com/nzfarmer1/mqtt2opcua
- cd mqtt2opcua
- npm link (view/edit examples/run.js)
- node examples/run.js

node creates an OPC server that:

1. Connects to an MQTT broker
2. Subscribes to '#' or a predefined set of topics
3. On 1st Message of an unseen topic it creates the folder paths and node (last part of path) with nodeId  s=topic
4. On all onMessage calls, it keeps track of the topic's payload
5. On Set requests it publishes the raw value from the OPCUA client
6. On Get requests it returns the formatted persistant payload


Note:

Recently upgraded to support MQTT auth and to remove spaces from OPCUA name in example

PLEASE REMEMBER TO CHECK FOR UPDATES REGULARLY!


<img src="mqtt2opcua.png"/>

Todo: Add authentication

For configuration options and formatting of payloads, see handlers example below.

<pre>

var  mqtt2opc = require("../mqtt2opcua").run;
var  Events = require('events').EventEmitter;

forward = new Events();
backward = new Events();

// Set up forward and reverse data conversion functions
// These are based on topic path - the finer grained pattern will be used.
// Examples below

forward.on("$SYS/broker/bytes/#", function(payload) {
    return {
            dataType: "Int32",
            value: parseInt(payload)
         }
});

backward.on("$SYS/broker/bytes/#", function(variant) {
            return {
                topic:variant.topic,
                payload:variant.value
            };
});

options = {
    opcHost:"localhost",
    opcPort:"4334",
    mqttHost:"localhost",
    mqttPort:"1883",
    debug:true,
    roundtrip:false,    // Set to true if you wish to enforce the integrity of round trip communications	
    forward:forward,	// data converter - mqtt -> opcua
    backward:backward,	// data converter - opcua -> mqtt
    //topics:['#','$SYS/#'] // Customize to override. These are the default so uncessary.
};

var server = new mqtt2opc(options);

</pre>

Please feel free to fork and improve.

<i>
Andrew McClure, Director <a href="http://agsense.co.nz">AgSense</a> -  Bringing I.o.T to NZ Agriculture
</i>

<b><a href="https://payment.swipehq.com/?product_id=EB82DA1340C7E">Koha (Donation)</a> - If you like our work and would like to make a small donation to help us continue</b>


CHANGELOG.txt
commit 2b02aa353b575fb692c6f6c8c45705c7a39cab1c
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 11:59:00 2015 +1200

    Initial commit

commit 55ad9227fc0c80ce1814a8daa0bd5e0d99eab84c
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 12:02:54 2015 +1200

    Tidy README

commit f3f78cd3b7a25c15dd388ba608799e7e6bcc7097
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 12:03:03 2015 +1200

    Tidy README

commit 9a6143d374ff537a9aa2c123ec130807703a5d0d
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 12:04:55 2015 +1200

    Tidy README

commit 20a4307981db20803e0f9400a1e74ac1ff6e6aaf
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 12:31:18 2015 +1200

    Tidy README

commit 3c0d07033a97b2ca55d66bd9e572d3d5235bf10f
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Wed Jul 22 22:31:36 2015 +1200

    improved logging

commit 5b3fc2d20ebe93623263335c4fda1bcba673e62f
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Thu Jul 23 19:35:22 2015 +1200

    removed merge dependency

commit 1304cd925027eb515594fe30f23a536894c421ba
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Thu Jul 23 21:06:17 2015 +1200

    Add ability to customise topic before publish

commit 45f2df29dfe2c2e194f859f4ed44c115d3317dcb
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Thu Jul 23 21:16:45 2015 +1200

    Add ability to customise topic before publish

commit b3f65389006b683da383bec807a21f837e8d9e37
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Mon Jul 27 10:30:36 2015 +1200

    bumped version

commit 9a954dfc2ba964b07297274e9ecfd03d6f7ee705
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Mon Jul 27 10:47:28 2015 +1200

    Improved error and OPC hostname handling

commit 7e9d195bd3a705a266042aad3faedcaa36420936
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Fri Jul 31 08:00:26 2015 +1200

    - Made debug a global function
    - Cast topic/payload from backward handlers toString
    - Added roundtrip option that optionally limits all updates to onMessage
    - Added opcName and roundtrip options to example script

commit 27f64417e40787aa86b1f74eac8afe0ce6361898
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Fri Jul 31 08:02:49 2015 +1200

    bumped version

commit b959f58dd317590239940074e9aa06cf320f47be
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Fri Jul 31 08:11:35 2015 +1200

    changed dependency name

commit a3099709cb93e3411cdfea3ba3b2969c6d7bff42
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Fri Jul 31 09:52:32 2015 +1200

    mqtt2opcua.js
    
    Add debug on Publish.  Fixed roundtrip mode

commit 8a4345ea5740132ec9bc3badfafad7e898799329
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Sun Aug 2 11:24:52 2015 +1200

    v0.0.13 lucky for some

commit ce7f91d30ec4f5960f71c9fa2e8fa51f6c58b24c
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Sun Aug 2 11:48:13 2015 +1200

    Updated README

commit 340571bc6d6f74fa468f22c4f5d9fecf275e1e02
Author: accyroy <accyroy@gmail.com>
Date:   Thu Oct 20 06:50:41 2016 -0400

    Ignition V7.9 cannot cope with space in opcName

commit c857656fe9842980fbdbb3f46d1075e8213b5a43
Author: accyroy <accyroy@gmail.com>
Date:   Thu Oct 20 07:03:36 2016 -0400

    Implemented addressSpace updates broken with node-opcua v0.0.48

commit 4b5fb989bebd26e0023401a3aeb389271b90c3a1
Author: accyroy <accyroy@gmail.com>
Date:   Thu Oct 20 13:26:23 2016 -0400

    Added authentication to the MQTT instance. Changed default payload to be a string.

commit 8aea07dcf7ab0d58f071edb0c750d21b8154b6bb
Merge: ce7f91d 4b5fb98
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Fri Oct 21 19:06:37 2016 +1300

    Merge pull request #2 from accyroy/master
    
    Fixed to work with V7.9 of Ignition and Latest version 0.0.57 of node-opcua

commit c614cf564c218d82441c18137700ed0dbe480efe
Author: hoblins <hoblins@users.noreply.github.com>
Date:   Sat Nov 26 10:59:24 2016 +0100

    Add files via upload

commit db025efc77b0500846b4191a93f799e4985bc31f
Merge: 8aea07d c614cf5
Author: hoblins <hoblins@users.noreply.github.com>
Date:   Sat Nov 26 11:02:08 2016 +0100

    Merge pull request #1 from hoblins/hoblins-patch-nodejs-6.9.1-7.2.0
    
    Add files via upload

commit 7d60505c958e3a9010228b761adf0bdfc7a6627f
Author: hoblins <hoblins@users.noreply.github.com>
Date:   Sat Nov 26 11:13:22 2016 +0100

    Fix for compatibility with nodejs 6.9.1 and 7.2.0
    
    Fixed problems with function .hasOwnProperty.
    "TypeError: this._events.hasOwnProperty is not a function"
    Not tested with older nodejs versions < 6.9.1.
    
    Ubuntu 16.10 64 bit
    Kernel 4.8.0-28-generic

commit 1659f933ad09b4f69d0aaeb0c45faef0dfc21fe1
Merge: 8aea07d 7d60505
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Sun Nov 27 05:38:53 2016 +1300

    Merge pull request #3 from hoblins/hoblins-patch-nodejs-6.9.1-7.2.0
    
    Hoblins patch for nodejs 6.9.1 and 7.2.0

commit a5d547930192634e716b4fb46ba9aad4b081395c
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Mon Nov 28 07:54:13 2016 +1300

    Updated package and added Change Log

commit 6d8079803e3328ebe6cb4b9fb05fb24513b9882a
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Mon Nov 28 07:55:05 2016 +1300

    Updated package and added Change Log

commit 74de31a84a81d5454616ca2f15cf3254a234a49f
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Thu Dec 15 16:54:32 2016 +1300

    Updated module definition for mqtt to suppler later versions of nodejs and mqtt

commit beb95f89f97311ae1865647840dae7779031b721
Author: Andrew McClure <andrew@southweb.co.nz>
Date:   Thu Dec 15 16:54:47 2016 +1300

    Updated module definition for mqtt to suppler later versions of nodejs and mqtt

commit 5a0bde50814852fab7df14242facc587a71ef127
Author: Andrew McClure <andrew@agsense.co.nz>
Date:   Fri Jan 14 15:18:40 2022 +1300

    Bumped version to 1.0.0


examples/run.js
var  mqtt2opc = require("../mqtt2opcua").run;
var  Events = require('events').EventEmitter;

forward = new Events();
backward = new Events();

// Set up forward and reverse data conversion functions
// These are based on topic path - the finer grained pattern will be used.
// Examples below

forward.on("$SYS/broker/bytes/#", function(payload) {
    return {
            dataType: "Int32",
            value: parseInt(payload)
         };
});

backward.on("$SYS/broker/bytes/#", function(variant) {
            return {
                topic:variant.topic,
                payload:variant.value
            };
});

options = {
    opcName:"MQTT_Local",
    opcHost:"localhost",
    opcPort:"4335",
    mqttHost:"localhost",
    mqttPort:"8324",
    mqttUsername:"opcua",
    mqttPassword:"secretpassword",
    debug:true,
    roundtrip:false,	// set to true to limit updates to onMessage (i.e. validate an accuator is set)
    forward:forward,	// data converter - mqtt -> opcua
    backward:backward,	// data converter - opcua -> mqtt
    //topics:['#','$SYS/broker/#'] // Customize to override. These are the default so uncessary.
};

var server = new mqtt2opc(options);




mqtt2opcua.js
"use strict";

var opcua = require("node-opcua"),
    mqtt = require('mqtt'),
    Events = require('events').EventEmitter,
    Qlobber = require('qlobber').Qlobber;


var  _debug  = false;



function debug(s) {
    if (_debug) {
        console.log(s);
    }
}


function Matcher(handlers) {
    Events.call(this);
    this.matcher = new Qlobber({
        separator: '/',
        wildcard_one: '+',
        wildcard_some: '#'
    });

    if (handlers.hasOwnProperty('_events')) {
        this._events = handlers._events;
    }
}

Matcher.prototype = Object.create(Events.prototype);
Matcher.prototype.match = function (topic) {
    if (Object.prototype.hasOwnProperty.call(this._events, topic)){
        return this._events[topic];
    }
    var matches = this.matcher.match(topic);
    return matches.length ? matches.pop() : null;
};

Matcher.prototype.init = function () {
    for (var e in this._events) {
        if (Object.prototype.hasOwnProperty.call(this._events, e)) {
            this.matcher.add(e, this._events[e]);
        }
    }
};

Matcher.prototype.hasDefault = function () {
    return this._events['#'] !== undefined;
};

var run = function (options) {

   _debug = options.debug || false;

 // Forward and backward handlers - see below for example
    var fhandlers = new Matcher(options.forward),
        bhandlers = new Matcher(options.backward);

    if (!fhandlers.hasDefault()) {
        fhandlers.on("#", function (payload) { // Default backward handler (MQTT -> OPCUA)
            return {
                dataType: "String",
                value: String(payload),
            };
        });
    }

    // Default backward handler
    if (!bhandlers.hasDefault()) {
        
        bhandlers.on("#", function (variant) { // Default backward handler (OPCUA -> MQTT)
            return {
                topic:variant.topic,
                payload:variant.value
            };
        });
    }

    fhandlers.init();
    bhandlers.init();


    // Let's create an instance of OPCUAServer

    var server = new opcua.OPCUAServer({
        hostname: options.opcHost || "127.0.0.1",
        port: options.opcPort || 4334, // the port of the listening socket of the server
        resourcePath: options.opcName || "UA/MQTT Bridge Server", // this path will be added to the endpoint resource name
        buildInfo : {
            productName: "MQTT2OPCUA Bridge",
            buildNumber: "0.1",
            buildDate: new Date(2015, 20, 7)
        }
    });


    function post_initialize() {

        debug("OPC Server Initialized");
        var persist = {},    // New persistent store
            nodes = {},
            paths = {},
            flagged = {};

        function onMessage(topic, payload) {

            debug("MQTT onMessage: " + topic + " (" + payload.length + " bytes)");

            var node,
                handler = fhandlers.match(topic);

            if (!handler) {
                console.error("No handler defined for " + topic);
                return;
            }

            if (nodes.hasOwnProperty(topic)) {
                persist[topic] = handler(payload).value; // Note: need to handle null responses
                flagged[topic] = false;
                return;
            }

            var sub,
                path = topic.split("/"),
                top  = (path[0] || path[1]); // Cater for paths starting with "/"

            node = paths[top] = (paths.hasOwnProperty(top)) ? paths[top] :
                                server.engine.addressSpace.addFolder("ObjectsFolder", { browseName: top});

            for (sub in path) {
                if (sub == 0 |!path.hasOwnProperty(sub)) {
                    continue;
                }

                top += "/" + path[sub];
                if (topic !== top) {
                    node = paths[top] = (paths.hasOwnProperty(top)) ? paths[top] :
                                        server.engine.addressSpace.addFolder(node, { browseName: path[sub]});
                } else {
                    debug("Creating folders for new topic: " + topic);
                    persist[topic] = handler(payload).value;
                    flagged[topic] = false;
                    try {
                        nodes[topic] = server.engine.addressSpace.addVariable({
                            componentOf: node,
                            browseName: path[sub],
                            path: topic,
                            nodeId: "s=" + topic,
                            dataType: handler(payload).dataType,
                            value: {
                                get: function () {
                                    debug("OPCUA Get: " + topic);
                                    return (!flagged[topic]) ?
                                        new opcua.Variant(handler(persist[topic])) :
                                            opcua.StatusCodes.BadCommunicationError;
                                    },
                                set: function (variant) {
                                    debug("OPCUA Set: " + topic);
                                    try {
                                        variant.topic=topic;
                                        var bhandler = bhandlers.match(topic),
                                            message  = bhandler(variant);
                                        if (message.hasOwnProperty("topic") &&
                                            message.hasOwnProperty("payload")) {
                                                if (!options.roundtrip)
                                                    persist[topic] = message.payload;
                                                else
                                                    flagged[topic] = true;
                                                debug("MQTT Publish: " + message.topic.toString() + " (" + message.payload.toString().substr(0,8) + ")" )
                                                server.mqtt.publish(message.topic.toString(),
                                                                    message.payload.toString());
                                            }
                                    } catch(e){
                                        console.error(e);
                                    }
                                    return opcua.StatusCodes.Good;
                                }
                            }
                        });
                    } catch (e) {
                        console.error(e);
                    }
                }
            }
            return;
        }


        server.start(function (err) {
	    if (err){
	       console.error(err);
	       process.exit(1);
	    }

	    var endpointUrl = server.endpoints[0].endpointDescriptions()[0].endpointUrl;
            console.log("OPC Server is now available at: " + endpointUrl +  " ( press CTRL+C to stop)");

            var mqttURL = 'mqtt://' + (options.mqttHost || "localhost") + ":" + (options.mqttPort || "1883");

            if (!(server.mqtt = mqtt.connect(mqttURL,{ username: options.mqttUsername, password: options.mqttPassword}))) {
                console.error("MQTT Unable to connect");
                process.exit(1);
            }

            server.mqtt.on('connect', function () {
                console.log("MQTT Connected to: %s\n", mqttURL);
                server.mqtt.subscribe(options.topics || ['$SYS/#', '#']);
            });

            server.mqtt.on('message', function (topic, message) {
                onMessage(topic, message);
            });

            server.mqtt.on('disconnect', function () {
                debug("MQTT Disconnected ");
                process.exit(1);
            });
        });
    }

    server.initialize(post_initialize);
};

module.exports.run = run;
